File: 0662-maximum-width-of-binary-tree/0662-maximum-width-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

import java.util.*;

class Solution {

    public class Pair {
        TreeNode node;
        long index;

        Pair(TreeNode node, long index) {
            this.node = node;
            this.index = index;
        }
    }

    public int widthOfBinaryTree(TreeNode root) {
        if (root == null) {
            return 0;
        }

        long maxWidth = 0;
        Queue<Pair> q = new LinkedList<>();
        q.offer(new Pair(root, 0));

        while (!q.isEmpty()) {
            int size = q.size();
            long levelStartIndex = q.peek().index;
            long first = 0, last = 0;

            for (int i = 0; i < size; i++) {
                Pair current = q.poll();
                long normalizedIndex = current.index - levelStartIndex;

                if (i == 0) {
                    first = normalizedIndex;
                }
                if (i == size - 1) {
                    last = normalizedIndex;
                }

                if (current.node.left != null) {
                    q.offer(new Pair(current.node.left, 2 * normalizedIndex + 1));
                }
                if (current.node.right != null) {
                    q.offer(new Pair(current.node.right, 2 * normalizedIndex + 2));
                }
            }

            maxWidth = Math.max(maxWidth, last - first + 1);
        }

        return (int) maxWidth;
    }
}

File: 0662-maximum-width-of-binary-tree/0662-maximum-width-of-binary-tree.cpp
class Solution {
public:
    int widthOfBinaryTree(TreeNode* root) {
        if (!root) return 0;
        queue<pair<TreeNode*, unsigned long long>> q;
        q.push({root, 0});
        int maxw = 0;
        help(q, maxw);
        return maxw;
    }

    void help(queue<pair<TreeNode*, unsigned long long>> q, int &maxw) {
        while (!q.empty()) {
            int n = q.size();
            unsigned long long left = q.front().second;
            unsigned long long right = left;
            for (int i = 0; i < n; i++) {
                auto [node, idx] = q.front(); q.pop();
                right = idx;
                if (node->left) {
                    q.push({node->left, 2 * idx});
                }
                if (node->right) {
                    q.push({node->right, 2 * idx + 1});
                }
            }
            maxw = max(maxw, (int)(right - left + 1));
        }
    }
};
